src/data/polynomial/coeff.lean

```diff
@@ -61,6 +62,47 @@ lemma coeff_mul (p q : polynomial R) (n : ℕ) :
   coeff (p * q) n = ∑ x in nat.antidiagonal n, coeff p x.1 * coeff q x.2 :=
 add_monoid_algebra.mul_apply_antidiagonal p q n _ (λ x, nat.mem_antidiagonal)
 
+lemma diag_inj
+  (n : ℕ) (x : ℕ × ℕ) (hx : x ∈ nat.antidiagonal n) (y : ℕ × ℕ) (hy : y ∈ nat.antidiagonal n) :
+  prod.fst x = prod.fst y → x = y :=
+begin
+  intro hg,
+  simp [nat.mem_antidiagonal] at *,
+  ext,
+  exact hg,
+  replace hg : x.fst = y.fst := by exact hg,
+  rw [← hx, hg] at hy,
+  rw add_left_cancel hy,
+end
+
+lemma coeff_mul' (n : ℕ) (p q : polynomial R) :
+  coeff (p * q) n = ∑ i in range (n + 1), coeff p i * coeff q (n - i) :=
+begin
+  rw coeff_mul,
```

Comment:
I think most of the proof below here could be replaced with `prod_antidiagonal_eq_prod_range_succ`